cypress-io/cypress
# feat: Run-all-specs-ui

- Closes #24468 and #24467: This closes all UI related to run-all-specs

### User facing changelog

Have introduced an run all specs under the experimental flag for e2e testing. This feature allows a user to run all specs under any directory or sub-directory.

When the experimental flag is enabled: hovering on a directory within the specs list or the inline spec list will trigger a action play button and text describing the number of specs within the hovered directory. A user would be able to click on the button to then run all specs within the  chosen directory. This applies for filtered specs as well.

### Additional details

- Have added a `RunAllSpecs.vue` component that is utilized in  `SpecsList.vue` and `InlineSpecsList.vue`. The component will accept a `specNumber` as a prop which represents the number of files within a directory.
- The most important is the `directoryChildren`. This is a reactive variable that is updated every time the "spec list tree" is re-built. It is a mapping `<{[key: string]: string[]}>`where the key is the relative path of a directory and the value is a list of specs within the directory. We use `directoryChildren[directoryId].length` to obtain the number of specs within a specific directory.

### Steps to test

There are two ways of testing the features.
A: Component Tests: `RunAllSpecs.cy.tsx`, `SpecsList.cy.tsx`, `InlineSpecsList.cy.tsx`
1. `cd packages/app`
2. Run `yarn cypress:open`
4. Choose component testing and test each of the files listed above. The contain run all specs related testing

B: Testing a project
1. Open two terminals
2. Run `yarn watch` in one terminal and wait for the app to be built.
3. Run `yarn cypress:open` in the second terminal after the completion of step 2
4.Select a project that contains `e2e` tests for example `cypress-example-kitchensink`. Go the project's `cypress.config` file and under the `e2e:{}` object add: `experimentalRunAllSpecs: true`
5. In the cypress app, select the project for which you have enabled run all specs
6. You should see the `Run N Specs` while hovering on directories in the specsList or see a play button when hovering over directories in the Inline specs list. When pressing these buttons you should be able to RUN ALL SPECS!


### How has the user experience changed?

Hovering over directory row:
<img width="747" alt="Screenshot 2022-11-18 at 8 35 40 PM" src="https://user-images.githubusercontent.com/65267668/202827905-ee00d33e-da16-467b-a0ec-95563fce3b62.png">
Hovering over button:
<img width="747" alt="Screenshot 2022-11-18 at 8 36 15 PM" src="https://user-images.githubusercontent.com/65267668/202827934-80de80f4-8a6e-4314-8999-edc7eca9d47b.png">
Hovering on Inline specs list directory row:
<img width="747" alt="Screenshot 2022-11-18 at 8 30 08 PM" src="https://user-images.githubusercontent.com/65267668/202827653-e8176e30-b565-4746-aa25-1b935c960dfb.png">
Hovering over button on inline specs list:
<img width="747" alt="Screenshot 2022-11-18 at 8 31 37 PM" src="https://user-images.githubusercontent.com/65267668/202827736-c785a20d-3d98-4ee5-a4be-1536c7ce2a03.png">

### PR Tasks


- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
